CMake/ConfigureCrossCompile.cmake

```diff
@@ -0,0 +1,33 @@
+if (CMAKE_CROSSCOMPILING)
+  include(board/flags-config.cmake)
+  if(NOT CMAKE_SYSROOT AND (NOT TOOLCHAIN_PREFIX))
+    message(FATAL_ERROR "Neither of CMAKE_SYSROOT or TOOLCHAIN_PREFIX is set, please set both of them and try again")
+  elseif(NOT CMAKE_SYSROOT)
+    message(FATAL_ERROR "Can not proceed CMAKE_SYSROOT is not set")
+  elseif(NOT TOOLCHAIN_PREFIX)
+    message(FATAL_ERROR "Cant not proceed TOOLCHAIN_PREFIXN is not set")
+  elseif(NOT OPENBLAS_TARGET)
+    message(FATAL_ERROR "Cant not proceed, board name is not set, please refer to documentation")
+  endif()
+endif()
+
+macro(search_openblas version)
+  set(BLA_STATIC ON)
+  find_package(BLAS)
+  if (NOT BLAS_FOUND OR (NOT BLAS_LIBRARIES))
+    get_deps(https://github.com/xianyi/OpenBLAS/releases/download/v${version}/OpenBLAS-${version}.tar.gz OpenBLAS OpenBLAS-${version}.tar.gz)
+    if (NOT MSVC)
+      if (NOT EXISTS "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
+        execute_process(COMMAND make TARGET=${OPENBLAS_TARGET} BINARY=${OPENBLAS_BINARY} HOSTCC=gcc CC=${CMAKE_C_COMPILER} FC=${CMAKE_FORTRAN_COMPILER} NO_SHARED=1
+                        WORKING_DIRECTORY ${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version})
+      endif()
+      file(GLOB OPENBLAS_LIBRARIES "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
+      set(BLAS_LIBRARIES ${OPENBLAS_LIBRARIES})
```

Comment:
I think you may also (or instead) need to set `BLAS_openblas_LIBRARIES` here.

Reply:
As you wish.

Reply:
I was remembering the name of the variable from memory incorrectly.  It is actually `BLAS_openblas_LIBRARY`, and when you set this, even when OpenBLAS is not installed in the buildroot environment, configuration and then build/linking succeeds.

I spent some time debugging and I think I have a conclusion but I have to go cook dinner now...  Here's what I found:

 * I am not sure what the case is with older CMake, but at least with newer CMake I don't need to set BLAS_LIBRARIES.  I think you should try removing it for older CMake and see what happens with the older version of FindBLAS.cmake.  (I forget at what CMake version this changed, but I think we wrote down the version number in a chat somewhere or something.)

 * I found that there is some strange interplay between `FindLAPACK.cmake` and `FindBLAS.cmake`.  It seems in many cases, `FindLAPACK.cmake` will call out to `FindBLAS.cmake`, and if a joint implementation of LAPACK/BLAS is found (which is the case for OpenBLAS), that will be used.  But in order to make the scripts find it, I had to set `LAPACK_openblas_LIBRARY` just like we set `BLAS_openblas_LIBRARY`.  I believe if you set those two variables (as well as `BLA_VENDOR` and `BLA_STATIC`), configuration should succeed on newer CMake.  (Older CMake might require `BLAS_LIBRARIES` and `LAPACK_LIBRARIES`... I'd try without those first, and if that fails, try adding them in?)

 * Once those changes above are done, you should be able to revert `FindArmadillo.cmake` to its original state in the master branch.

I think that's it... let me know what you find out if you try it.  For now it's dinner time. :)

Reply:
Each test done after cleaning caching, here is my matrix build:
Localy
set only `BLAS_LIBRARIES` does not work
set only `BLAS_openblas_LIBRARY` works
set `BLAS_openblas_LIBRARY` and `BLAS_LIBRARIES` works (thanks to the `BLAS_openblas_LIBRARY`)
set only `LAPACK_openblas_LIBRARY` does not work
set `BLAS_openblas_LIBRARY` and `LAPACK_openblas_LIBRARY` works (thanks to the `BLAS_openblas_LIBRARY`)
set `BLAS_LIBRARIES` and `LAPACK_LIBRARIES` does not work
on CI:
set only `BLAS_LIBRARIES` does not work
set only `BLAS_openblas_LIBRARY` does not work
set `BLAS_openblas_LIBRARY` and `BLAS_LIBRARIES` does not work
set only `LAPACK_openblas_LIBRARY` does not work
set `BLAS_openblas_LIBRARY` and `LAPACK_openblas_LIBRARY`  does not work
set `BLAS_LIBRARIES` and `LAPACK_LIBRARIES` does not work

Maybe it is the time for a new version of CMake on CI from the current one 3.13.4?

Reply:
CMake 3.13.4 (and newer versions that still have the old FindBLAS/FindLAPACK) will still be present in the wild, so we can't upgrade.  We will have to find a solution.  I guess I will dig further tomorrow.

Reply:
Using CMake version 3.10.2 seems to work by setting `BLAS_openblas_LIBRARY` and `LAPACK_openblas_LIBRARY`.

Reply:
Indeed I was even able to use `BLAS_openblas_LIBRARY` and `LAPACK_openblas_LIBRARY` successfully on CMake 3.13.4.  I simply uncommented the lines setting those variables in `ConfigureCrossCompile.cmake`:

```
root@0a5042e2b194:/src/mlpack/build-arm64# git diff
diff --git a/CMake/ConfigureCrossCompile.cmake b/CMake/ConfigureCrossCompile.cmake
index 3078d400d..d0f1fa1b0 100644
--- a/CMake/ConfigureCrossCompile.cmake
+++ b/CMake/ConfigureCrossCompile.cmake
@@ -31,8 +31,8 @@ macro(search_openblas version)
       file(GLOB OPENBLAS_LIBRARIES "${CMAKE_BINARY_DIR}/deps/OpenBLAS-${version}/libopenblas.a")
       set(BLAS_LIBRARIES ${OPENBLAS_LIBRARIES})
       set(LAPACK_LIBRARIES ${OPENBLAS_LIBRARIES})
-      # set(BLAS_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
-      # set(LAPACK_openblas_LIBRARY ${OPENBLAS_LIBRARIES})  
+      set(BLAS_openblas_LIBRARY ${OPENBLAS_LIBRARIES})
+      set(LAPACK_openblas_LIBRARY ${OPENBLAS_LIBRARIES})  
       message(STATUS "SHOW BLAS libraries: ${BLAS_LIBRARIES}")
       set(BLA_VENDOR OpenBLAS)
       set(BLAS_FOUND ON)
```

I suspect that `BLAS_LIBRARIES` and `LAPACK_LIBRARIES` don't need to be set.  Can you double-check to see if you see the same results on your setup?